steps 8 and 9

jchavarri opened the pull request:
- Step 8: Using input fields to get data for any user
- Step 9: Validating data with abstract modules, handling edge cases and errors

jchavarri committed:
steps 8 and 9

feihong commented on README.md:
```diff
@@ -629,6 +629,151 @@ Can you try updating the rendering function to also show `entry.content` and
+To do this, we will have to update our app in a few ways.
+
+First of all, we will add a couple more state hooks to store `username` and the
+`debounceTimeout`. For this, we will use `React.useState`, which we have used
```
It might help to mention that the initial value of `debounceTimeout` is `None` and that its type is `option(Js.Global.timeoutId)`

jchavarri replied:
There's no more debounce.

feihong commented on README.md:
```diff
@@ -629,6 +629,151 @@ Can you try updating the rendering function to also show `entry.content` and
+    - set a new timeout using
+      [Js.Global.setTimeout](https://melange.re/v4.0.0/api/re/melange/Js/Global/#val-setTimeout).
+      Let's set it to 500ms.
+    - store the new timeout id using `setDebounceTimeout` function from the
```
I'm not sure about adding username and timeout in the same step. It would be easier if this step focused on adding username and just refetch when you hit Enter in the input.

jchavarri replied:
Good idea. I will remove the debounce and use `onKeyDown` instead.

jchavarri replied:
Actually, removing this debounce leads to more complicated logic in the next step, because we have to account for more states that weren't there before, like the user is typing but hasn't pressed `Enter` yet. I am not sure which path is easier.

jchavarri replied:
Ended up keeping the `Enter` approach, see https://github.com/ahrefs/react-alicante-workshop/pull/6/commits/1e63afcbe33c61d6e38e696b507d10400a80f14b.

feihong commented on README.md:
```diff
@@ -629,6 +629,151 @@ Can you try updating the rendering function to also show `entry.content` and
+Finally, we have to modify our data fetching logic to get data for the current
+user. This will involve the following steps:
+
+- Remove the previous effect and move its content to a new function `fetchFeed`.
```
Maybe mention that `fetchFeed` should be a nested function inside `App.make`

jchavarri replied:
It's not strictly necessary, one could move the function outside and pass the username + callbacks as params, right?

feihong replied:
I probably `fetchFeed` a nested function from force of habit

feihong commented on README.md:
```diff
@@ -629,6 +629,151 @@ Can you try updating the rendering function to also show `entry.content` and
+    - store the new timeout id using `setDebounceTimeout` function from the
+      `useState` hook
+
+### Step 8 completion check
```
I feel like more hints are needed. For example, I flailed for a while until I realized that I needed to move `setData(_ => Loading);` inside the `fetchFeed` function.

jchavarri replied:
This is not strictly necessary. In my solution, I was adding the `setData(_ => Loading)` outside that function. i think it's ok to leave some room for folks to experiment with different solutions.

feihong commented on README.md:
```diff
@@ -629,6 +629,151 @@ Can you try updating the rendering function to also show `entry.content` and
+large already, let's create a new module `Username.re`. In this module, we will
+define the following:
+
+- `type t = string`
```
```suggestion
- `type t = string;`
```

feihong commented on README.md:
````diff
@@ -629,6 +629,151 @@ Can you try updating the rendering function to also show `entry.content` and
+    let validatedUsername =
+      React.useMemo1(() => Username.make(username), [|username|]);
+```
+- Make sure that `fetchFeed` takes a value `Username.t` rather than a plain
````
I think the part where it says to make `fetchFeed` take username as an argument should be a bit more explicit, I totally missed that and made `fetchFeed` an nested function which doesn't take any arguments.

jchavarri replied:
I added more explicit instructions in https://github.com/ahrefs/react-alicante-workshop/pull/6/commits/1e63afcbe33c61d6e38e696b507d10400a80f14b.

feihong commented on README.md:
```diff
@@ -629,6 +629,151 @@ Can you try updating the rendering function to also show `entry.content` and
+
+> [!TIP]
+> 
+> You can use either `if` / `else` or a `switch` statement for this.
```
```suggestion
> You can use either `if` / `else` or a `switch` expression for this.
```

feihong commented on README.md:
```diff
@@ -629,6 +629,151 @@ Can you try updating the rendering function to also show `entry.content` and
+- Make sure that `fetchFeed` takes a value `Username.t` rather than a plain
+  string
+- Adapt the effect to depend on `validatedUsername`, not `username`
+- Modify the render function to pattern match on both `validatedUsername` and
```
I don't understand this part. In my code, I put a switch expression after the `input`:

```reason
           <input
             id="username-input"
             value=username
             onChange={event => {
               setUsername(event->React.Event.Form.target##value)
             }}
             placeholder="Enter GitHub username"
           />
           {switch (validatedUsername) {
            | Ok(_) => React.null
            | Error () => <div> {React.string("Invalid username")} </div>
            }}
```
This seems sufficient? I didn't touch the switch expression that displays the feed.

jchavarri replied:
I prefer to handle all rendering logic in the same switch. The downside of your suggested approach is that one might end up with the UI showing both "Invalid username" and "Loading" messages, for example, which can be quite confusing from the user perpective.

jchavarri replied:
I ended up modeling this like a state machine, which I think makes the code easier to follow (no more `validatedUsername`), and showcases better the power of variants, see https://github.com/ahrefs/react-alicante-workshop/pull/6/commits/1e63afcbe33c61d6e38e696b507d10400a80f14b. lmk what you think.

feihong commented on README.md:
```diff
@@ -629,6 +629,151 @@ Can you try updating the rendering function to also show `entry.content` and
+message. Or we can type an invalid username like `invalid-` to see the invalid
+username message.
+
+Finally, we can type an inexistent user like `a---1` to see the invalid response
```
```suggestion
Finally, we can type an nonexistent user like `a---1` to see the invalid response
```

feihong commented on README.md:
```diff
@@ -629,6 +629,151 @@ Can you try updating the rendering function to also show `entry.content` and
+
+For this, we will have to modify `App` component in a couple of places.
+
+The `fetchFeed` function should check for the response status using
```
I'm a little ashamed to admit I struggled with this part. I had the general idea of what to do, but it took a while for me to figure out that I needed to add `resolve` and `ignore` calls. Maybe it'll be less of an issue for participants since they won't be used to having `Promise.each`?

jchavarri replied:
I added some snippet to showcase how this modification should happen in https://github.com/ahrefs/react-alicante-workshop/pull/6/commits/1e63afcbe33c61d6e38e696b507d10400a80f14b.

feihong reviewed (approved):
Styles are still broken. For some reason, I can't scroll all the way to the top of the feed:

<img width="1069" alt="image" src="https://github.com/user-attachments/assets/4d6d56ae-cade-4ee9-b102-937766cfee21">

For some reason, it just cuts off everything above a certain point.

---

The output of the API looks off. For example, all links titles seem to be empty:

```
            "links": [
                {
                    "title": "",
                    "href": "https://github.com/ahrefs/react-alicante-workshop/compare/38e431c991...9903250b4d"
                }
            ],
```

I checked several and they all had empty link titles.

---

I didn't notice until now but the structure of App.re is a bit weird, with the inner module also being called App. Would it make sense to create an Index.re file?

---

It took me longer than I expected to get through steps 8 and 9. I think this will be a pretty challenging workshop.

jchavarri committed:
english

Co-authored-by: Feihong Hsu <feihong.hsu@gmail.com>

jchavarri committed:
english

Co-authored-by: Feihong Hsu <feihong.hsu@gmail.com>

davesnx commented on README.md:
```diff
@@ -629,6 +629,151 @@ Can you try updating the rendering function to also show `entry.content` and
+In this step, we'll enhance our app by letting users enter a GitHub username to
+fetch and display their activity feed.
+
+To do this, we will have to update our app in a few ways.
```
Worth adding the "product" idea?

To do this, we will have to update our app in a few ways: `add an input for the username, store the username, fire the request to Github and react will display the results`

jchavarri replied:
added in https://github.com/ahrefs/react-alicante-workshop/pull/6/commits/1e63afcbe33c61d6e38e696b507d10400a80f14b.

davesnx commented on README.md:
```diff
@@ -629,6 +629,151 @@ Can you try updating the rendering function to also show `entry.content` and
+Note how we are leaving `t` without a definition, this makes the type abstract.
+This way, every time we need to use a GitHub username in our app, we can use
+`Username.t` as a type, and this will guarantee that the string has gone through
+the validation before being used in the function.
```
I would explain verbally the difference between TypeScript and OCaml regarding nominal and structural. Those details are hard to grasp without an example, but this is a great example to showcase.

davesnx replied:
also, can explain that running a regex might be silly and you might want to run this "validation" async by hitting their API

jchavarri replied:
> I would explain verbally the difference between TypeScript and OCaml regarding nominal and structural. Those details are hard to grasp without an example, but this is a great example to showcase.

I don't see how nominal types enable abstract types 🤔 it's rather the separation between interface and implementation no? see e.g. https://github.com/microsoft/TypeScript/issues/15408#issuecomment-298162519. Maybe i don't understand fully, can you elaborate?

> can explain that running a regex might be silly and you might want to run this "validation" async by hitting their API

I can't find any GitHub API endpoints to validate a username, do you know how?

jchavarri committed:
fix syntax

Co-authored-by: Feihong Hsu <feihong.hsu@gmail.com>

jchavarri committed:
modification after review

jchavarri commented:
> For some reason, it just cuts off everything above a certain point.

This was an issue on the global CSS, which was trying to center vertically. It's been removed from `main` now: https://github.com/ahrefs/react-alicante-workshop/commit/fea341416260edd055739fd1c92a6168b155fb5f.

> The output of the API looks off. For example, all links titles seem to be empty:

Yeah, I noticed too. We might just have to remove any references to `links` from the workshop.

> I didn't notice until now but the structure of App.re is a bit weird, with the inner module also being called App. Would it make sense to create an Index.re file?

I will try to remember to change it.

> It took me longer than I expected to get through steps 8 and 9. I think this will be a pretty challenging workshop.

I agree the initial version that you reviewed was a bit "rough", but I still think it's useful to get to a point where attendants "sweat" a little bit to reach a solution. After your feedback, I have added extensive clarifications, simplified the code and included more snippets to make these steps a bit more manageable, I would appreciate if you could take it for another spin and see if things are easier to follow now? Thanks!

feihong commented:
> Yeah, I noticed too. We might just have to remove any references to links from the workshop.

The other thing I noticed is that even though `links` is an array, there's only ever one item in it for every case I've seen. So it's not very interesting in terms of structure.

feihong commented on README.md:
```diff
@@ -629,6 +629,274 @@ Can you try updating the rendering function to also show `entry.content` and
+    onChange={event => {
+      setUsername(event->React.Event.Form.target##value)
+    }}
+    onKeyDown={event =>
```
This chunk of code is missing a pair of parentheses

jchavarri replied:
https://github.com/ahrefs/react-alicante-workshop/pull/6/commits/6d2eda65458312c280fbbc53c94949d4d37b30e7

feihong commented on README.md:
```diff
@@ -629,6 +629,274 @@ Can you try updating the rendering function to also show `entry.content` and
+
+Let's modify `fetchFeed` to take a value of type `Username.t` rather than
+`string`, so we make sure there are no requests being done with invalid
+usernames. This essentially consists on replaing `username` with
```
```suggestion
usernames. This essentially means replacing `username` with
```

jchavarri replied:
https://github.com/ahrefs/react-alicante-workshop/pull/6/commits/6d2eda65458312c280fbbc53c94949d4d37b30e7

feihong commented on README.md:
````diff
@@ -629,6 +629,274 @@ Can you try updating the rendering function to also show `entry.content` and
+    | Loaded(Error(msg)) => <div> {React.string(msg)} </div>
+    | Loaded(Ok(feed)) => ...
+  }}
+```
````
After following along, there's one thing that was still not compiling, though it's pretty minor and not hard to fix: Changing `input`'s `value` prop from `value=username` to `value={state.username}`.

jchavarri replied:
https://github.com/ahrefs/react-alicante-workshop/pull/6/commits/6d2eda65458312c280fbbc53c94949d4d37b30e7

feihong commented on README.md:
```diff
@@ -629,6 +629,274 @@ Can you try updating the rendering function to also show `entry.content` and
+        |> P.resolve;
+      };
+    })
+|> ignore;
```
I think this template helps a lot but I still ended up forgetting to place a `P.resolve`. Incidentally, one thing I noticed is the error message you see when you use module alias was a little easier to understand than the one you get if you use local open.

Module alias:

```
File "src/App.re", line 69, characters 18-68:
69 |                   setState(state => {...state, step: Loaded(data)});
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: This expression has type unit but an expression was expected of type
         'a P.t
```

(maybe this is a case for changing module alias to `module Promise = Js.Promise;`, in case they forget what P is aliased to)

Local open:

```
File "src/App.re", line 70, characters 18-68:
70 |                   setState(state => {...state, step: Loaded(data)});
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: This expression has type unit but an expression was expected of type
         'a t
```

feihong reviewed (approved):
Thanks for the changes, I think it was significantly easier working through it this time!

I noticed that the useEffect0 code snippet of [section 7.1](https://github.com/ahrefs/react-alicante-workshop/tree/step-8-n-9?tab=readme-ov-file#71-decoding-the-fetched-data) you use a local open instead of the module alias for Js.Promise. I think that could be confusing without more explanation of local open. 

But also, I think we should avoid local open entirely for this workshop because getting errors about type `'a t` seems rough for beginners.

I made a repo for the code I wrote so far: https://github.com/feihong/react-alicante-workshop

jchavarri committed:
fixes

jchavarri commented:
> I noticed that the useEffect0 code snippet of [section 7.1](https://github.com/ahrefs/react-alicante-workshop/tree/step-8-n-9?tab=readme-ov-file#71-decoding-the-fetched-data) you use a local open instead of the module alias for Js.Promise. I think that could be confusing without more explanation of local open.
> 
> But also, I think we should avoid local open entirely for this workshop because getting errors about type `'a t` seems rough for beginners.

I agree, @davesnx also suggested to get rid of local opens, so I'll do that in a follow up PR.
